src/somef/extract_software_type.py

```diff
@@ -0,0 +1,325 @@
+                            readme_contents = readme_file.read()
+                            
+                            title_words = title.split()
+
+                            pattern = r'([^.?!]*(?:\b|\W){}(?:\b|\W)[^.?!]*[.?!])'.format('|'.join(map(re.escape, title_words)))
+                            sentences = re.findall(pattern, readme_contents, flags=re.IGNORECASE)
+                            for sentence in sentences:
+                                if re.search(rf'\b{title}\b', sentence, flags=re.IGNORECASE) and re.search(r'\b(pipeline|workflow)\b', sentence, flags=re.IGNORECASE):
+                                    return True
+
+                            pattern_md=r'##.*\b(workflow|pipeline)\b'
+                            pattern_rst=r'^([^=\n]+(?:\n(?![-=]).*)*\b(workflow|pipeline)\b(?:\n(?![-=]).*)*)\n=+'
+                            
+                            match_md=re.findall(pattern_md,readme_contents,re.IGNORECASE)
+                            match_rst=re.findall(pattern_rst,readme_contents, re.MULTILINE | re.IGNORECASE)
+                            
+                            if match_md or match_rst:
+                                return True
+
+                    except:
+                        continue
+
+            if file.endswith(extensions) or file =="Snakefile":
+                total_workflows+=1
+                file_path = os.path.join(root, file)
+                if is_file_workflow(file_path):
+                    list.append(file_path)
+                    good_workflows+=1
+                else:
+                    continue
+            if check_name(file)==True:
+                return True
+    if list!=[]:
+        return True
+    else:
+        return False
+
+def check_name(filename):
+    """Assisting function for check_workflow to look for specific named files"""
+    pattern1 = re.compile(r"pipeline", re.IGNORECASE)
+    pattern2=re.compile(r"(model|example)",re.IGNORECASE)
+    programming_language_extensions = (".py", ".java", ".cpp", ".c", ".php", ".rb", ".js", ".html", ".css", ".go", ".swift", ".scala", ".pl",".ipynb")
+    if pattern1.search(filename) and pattern2.search(filename):
+        if filename.endswith(programming_language_extensions):
+            return True
+    elif pattern1.search(filename) and not pattern2.search(filename):
+        return False
+    elif (not pattern1.search(filename) and not pattern2.search(filename)) or (not pattern1.search(filename) and pattern2.search(filename)):
+        return False
+
+def is_notebook_code(file_path):
+    has_code = False
+    num_code_cells = 0
+    num_total_cells = 0
+    nb = nbformat.read(file_path, as_version=4)
+    for cell in nb['cells']:
+        if cell['cell_type'] == 'code':
+            num_total_cells += 1
+            if cell['source'].strip():
+                num_code_cells += 1
+                has_code = True
+    if has_code and num_code_cells / num_total_cells >= 0.55:
```

Comment:
This number looks arbitrary. Please document why this is this way in a comment and provide evidence to support this decision